### DISCUSSION TITLE
How can I change Agent.pos ?

### DISCUSSION THREAD

**Discussion by Nebukam** [Nebukam/com.nebukam.orca#5] [Q&A] [answered]
*Originally asked through this [issue](https://github.com/Nebukam/com.nebukam.orca/issues/4) by @chasing2moro*

I change the value pos of Agent, but It does not work. Because after I do that, Agent.pos is assigned by AgentData.position immediately. So please help me： How can I change Agent.pos ?

**Comment by Nebukam** [comment]
`agent.pos` updating the `AgentData.position` is the expected behavior; `pos` is used to set the current position of the agent inside the simulation.

If you're trying to set a *goal* location for the agent, this is something you need to implement yourself: ORCA is not a pathfinding solution and the simulation works by using the agent *actual position* (`agent.loc`) combined with a *desired velocity* (`agent.prefVelocity`). 
From there you run the simulation and the position of the agent will be updated according to its own velocity & other agents inside that same simulation.

There are samples available with the package, that can be installed through the package manager; I recommend looking at **Setup example (flock)**
![image](https://github.com/Nebukam/com.nebukam.orca/assets/5935407/3fb364d9-aa0a-49b2-bbcf-0db226d747ab)

Let me know if this helps!

**Reply by chasing2moro** [reply]
100 thx for your enthusiastic answer

**Reply by chasing2moro** [reply]
After I practice Setup example (flock). It still need several frames time (≈ 0.1s) to reach the goal location. It's there any solution to reach a goal location immediately？

**Reply by Nebukam** [reply]
If you don't want to simulate the movement toward the goal then setting `agent.pos` as you did initially would be the way to go.

Can you provide a bit more context on what you're trying to achieve?

**Reply by chasing2moro** [reply]
I  implement scenario : Teleport a unit from one postion to other position immediately, and than the unit continue to walk to enemy.

**Reply by Nebukam** [reply, marked as answer]
The easiest way to go about that would be something along the lines of:

1. Runs every frame:
```CSharp
// Given two agents, A will move toward B
agentA, agentB (enemy)

// AgentB being a moving goal, make sure AgentA prefVelocity is always pointing toward it
agentA.prefVelocity = (normalized direction from agentA to agentB) * movement speed

// Only run the simulation after that, ensuring agents have the most up-to-date 'intentions'
simulation.Run(Time.deltaTime); // Run instead of Schedule
```

2. Runs once/when needed, but make sure it's applied before running a new simulation step
```CSharp
// Teleport agentA at a desired location
agentA.pos = teleport location
```

In that scenario, since agentA prefVelocity is kept up-to-date each frame, it should always move toward agentB no matter its position.
*Note that if you teleport an agent immediately after the simulation has run (as opposed to right before), it will still take one step of simulation to adjust agentA' orientation*

Since this is multi-threaded code, as soon as a new simulation step is ran, it will take a snapshot of the current `agent.pos` and `agent.prefVelocity` to do the maths. **What may be happening is that you update `agent.pos` while the simulation step is not done running yet, and once the step is finished it overrides the position you set manually with the simulation result.**

In order to make the simulation code run synchronously (i.e without multithreading) you can call `simulation.Run(Time.deltaTime)` instead of `simulation.Schedule(Time.deltaTime)` -- this will force the simulation to run on the main thread and can be helpful for debugging behaviors related to execution order

**Reply by chasing2moro** [reply]
Thank you very much. I show my appreciation to your sincerely care about such a stranger me. You light up my path to solve the key problem.

**Reply by chasing2moro** [reply]
I finally got a solution that move to a goal postion immediately within 0.0001s and also without calculating Agent.prefVelocity-vector:

1. Invoke ORCA.TryComplete function 
2. Assign goal pos to Agent.pos 
3. Invoke ORCA.Schedule function. This will lead to `AgentData.position` assigned by Agent.pos , And then Agent.pos is assigned back by  `AgentData.position`

```
        ORCA.TryComplete();

        Agent.pos = #goal pos#

        ORCA.Schedule(Time.deltaTime);
```
